src/lib/sw_engine/tvgSwFill.cpp

```diff
@@ -70,31 +70,38 @@ static bool _updateColorTable(SwFill* fill, const Fill* fdata, const SwSurface*
         auto curr = colors + j;
         auto next = curr + 1;
         auto delta = 1.0f / (next->offset - curr->offset);
-        a = (next->a * opacity) / 255;
+        auto a = (next->a * opacity) / 255;
         if (!fill->translucent && a < 255) fill->translucent = true;
 
-        auto r = ALPHA_MULTIPLY(next->r, a);
-        auto g = ALPHA_MULTIPLY(next->g, a);
-        auto b = ALPHA_MULTIPLY(next->b, a);
-
-        auto rgba2 = surface->blender.join(r, g, b, a);
+        auto rgba2 = surface->blender.join(next->r, next->g, next->b, a);
 
         while (pos < next->offset && i < GRADIENT_STOP_SIZE) {
             auto t = (pos - curr->offset) * delta;
             auto dist = static_cast<int32_t>(256 * t);
             auto dist2 = 256 - dist;
-            fill->ctable[i] = COLOR_INTERPOLATE(rgba, dist2, rgba2, dist);
+
+            auto color = COLOR_INTERPOLATE(rgba, dist2, rgba2, dist);
+            uint8_t a = color >> 24;
+            uint8_t c1 = ALPHA_MULTIPLY((color >> 16) & 0xff, a);
+            uint8_t c2 = ALPHA_MULTIPLY((color >> 8) & 0xff, a);
+            uint8_t c3 = ALPHA_MULTIPLY(color & 0xff, a);
+            fill->ctable[i] = (a << 24 | c1 << 16 | c2 << 8 | c3);
+            
             ++i;
             pos += inc;
         }
         rgba = rgba2;
     }
 
+    r = ALPHA_MULTIPLY(r, a);
+    g = ALPHA_MULTIPLY(g, a);
+    b = ALPHA_MULTIPLY(b, a);
+
     for (; i < GRADIENT_STOP_SIZE; ++i)
-        fill->ctable[i] = rgba;
```

Comment:
I think you can pre-compute rgba and use it in the loop.

Reply:
of course, but what's worse, I was using the wrong values (from the first stop instead from the last one) ! fixed now :)